samza-api/src/main/java/org/apache/samza/table/TableDescriptorsFactory.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.table;
+
+import java.util.List;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.operators.TableDescriptor;
+
+
+/**
+ * Factory to create a list of {@link TableDescriptor} objects to describe one or more Samza tables. This is the
+ * mechanism for providing table support for Samza low level API.
+ *
+ * Developers writing Samza jobs using Samza table(s) should describe the table(s) by implementing
+ * TableDescriptorFactory in their task. Please note that the class that implements this factory should also implement
+ * one of {@link org.apache.samza.task.StreamTask} or {@link org.apache.samza.task.AsyncStreamTask}.
+ *
+ * NOTE: {@link TableDescriptorsFactory} is instantiated in the config rewriter to get the table descriptors. Hence,
+ * please DO NOT access any task instance member variables in the getTableDescriptors API.
+ *
+ * Typical user code using Samza tables should look like the following:
+ *
+ * <pre>
+ * {@code
+ * public class SampleTask implements TableDescriptorsFactory, InitableTask, StreamTask {
+ *   private ReadableTable<String, Long> remoteTable;
+ *   private ReadWriteTable<String, String> localTable;
+ *
+ *   {@code @Override}
+ *   public List<TableDescriptor> getTableDescriptors() {
+ *     List<TableDescriptor> tableDescriptors = new ArrayList<>();
+ *     final TableReadFunction readRemoteTable = (TableReadFunction) key -> null;
+ *     tableDescriptors.add(new RemoteTableDescriptor<>("remote-table-1")
+ *       .withReadFunction(readRemoteTable)
+ *       .withSerde(KVSerde.of(new StringSerde(), new StringSerde())));
+ *
+ *     tableDescriptors.add(new RocksDbTableDescriptor("local-table-1")
+ *       .withBlockSize(4096)
+ *       .withSerde(KVSerde.of(new LongSerde(), new StringSerde<>())));
+ *       .withConfig("some-key", "some-value");
+ *     return tableDescriptors;
+ *   }
+ *
+ *   {@code @Override}
+ *   public void init(Config config, TaskContext context) {
+ *     remoteTable = (ReadableTable<String, String>) context.getTable(“remote-table-1”);
```

Comment:
nit: given the process() method below, maybe we should call this remote-profile-table, and the local-profile-table to match up. Otherwise, below we should just use generic names, eg. some-field, instead of memberId.